Source/Core/VideoCommon/PerformanceMetrics.cpp

```diff
@@ -0,0 +1,254 @@
+  m_cpu_dt = std::max(cpu_now - m_cpu_time, DT::zero());
+  m_cpu_time = cpu_now;
+
+  const double speed = DT_s(m_cpu_dt) / DT_s(m_real_dt);
+  const double a = 1.0 - std::exp(-(std::max(m_cpu_dt, m_real_dt) / DT_s(2.0)));
+
+  if (std::isfinite(m_emulation_speed))
+    m_emulation_speed += a * (speed - m_emulation_speed);
+  else
+    m_emulation_speed = speed;
+}
+
+double PerformanceMetrics::GetFPS() const
+{
+  return m_fps_counter.GetHzAvg();
+}
+
+double PerformanceMetrics::GetVPS() const
+{
+  return m_vps_counter.GetHzAvg();
+}
+
+double PerformanceMetrics::GetSpeed() const
+{
+  return 100.0 * m_speed_counter.GetHzAvg() / VideoInterface::GetTargetRefreshRate();
+}
+
+double PerformanceMetrics::GetLastSpeedDenominator() const
+{
+  return DT_s(m_speed_counter.GetLastRawDt()).count() * VideoInterface::GetTargetRefreshRate();
+}
+
+double PerformanceMetrics::GetEstimatedEmulationSpeed() const
+{
+  return m_emulation_speed;
+}
+
+void PerformanceMetrics::DrawImGuiStats(const float backbuffer_scale) const
+{
+  const float bg_alpha = 0.7f;
+  const auto imgui_flags = ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoInputs |
+                           ImGuiWindowFlags_NoMove | ImGuiWindowFlags_NoSavedSettings |
+                           ImGuiWindowFlags_NoScrollbar | ImGuiWindowFlags_NoNav |
+                           ImGuiWindowFlags_AlwaysAutoResize | ImGuiWindowFlags_NoFocusOnAppearing;
+
+  const double fps = GetFPS();
+  const double vps = GetVPS();
+  const double speed = GetSpeed();
+
+  // Change Color based on % Speed
+  float r = 0.0f, g = 1.0f, b = 1.0f;
+  if (g_ActiveConfig.bShowSpeedColors)
+  {
+    r = 1.0 - (speed - 80.0) / 20.0;
+    g = speed / 80.0;
+    b = (speed - 90.0) / 10.0;
+  }
+
+  const float window_padding = 8.f * backbuffer_scale;
+  const float window_width = 93.f * backbuffer_scale;
+  float window_y = window_padding;
+  float window_x = ImGui::GetIO().DisplaySize.x - window_padding;
+
+  const float graph_width = 50.f * backbuffer_scale + 3.f * window_width + 2.f * window_padding;
+  const float graph_height =
+      std::min(200.f * backbuffer_scale, ImGui::GetIO().DisplaySize.y - 85.f * backbuffer_scale);
+
+  ImGui::PushStyleVar(ImGuiStyleVar_WindowBorderSize, 0.f);
+  ImGui::PushStyleVar(ImGuiStyleVar_WindowRounding, 14.f * backbuffer_scale);
+  if (g_ActiveConfig.bShowGraphs)
+  {
+    ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, ImVec2(0, 4.f * backbuffer_scale));
+
+    // Position in the top-right corner of the screen.
+    ImGui::SetNextWindowPos(ImVec2(window_x, window_y), ImGuiCond_Always, ImVec2(1.0f, 0.0f));
+    ImGui::SetNextWindowSize(ImVec2(graph_width, graph_height));
+    ImGui::SetNextWindowBgAlpha(bg_alpha);
+    window_y += graph_height + window_padding;
+
+    if (ImGui::Begin("PerformanceGraphs", nullptr, imgui_flags))
+    {
+      const static int num_ticks = 17;
+      const static double tick_marks[num_ticks] = {0.0,
```

Comment:
preferably

```suggestion
      static constexpr std::array<double, num_ticks> tick_marks = {0.0,
```

since std arrays can have bounds checks explicitly enabled in implementations through defines, whereas C arrays cannot